### PR TITLE
Adding IBLevelSet/ex6: floating rectangular block

### PR DESCRIPTION
**Pull request by Kaustubh-Khedkar**
### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [ ] Was clang-format (i.e., `make indent`) run? For more information see
      `scripts/formatting/README.md`.
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [ ] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

File: examples/IBLevelSet/ex6/LSLocateGasInterface.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+namespace IBTK
+{
+class HierarchyMathOps;
+}
+
+/*
+ * Pre processing call back function to be hooked into IBAMR:LInitStrategy
+ *
+ * \TODO: Let's move this out of the global namespace and use "snake case" for static function names.
```
Comment (boyceg) [inline examples/IBLevelSet/ex6/LSLocateGasInterface.h:35]: Could you address this TODO in this example?

File: examples/IBLevelSet/ex6/LevelSetInitialCondition.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    /*!
+     * \brief Evaluate the function on the patch interior.
+     */
+    void setDataOnPatch(const int data_idx,
+                        Pointer<SAMRAI::hier::Variable<NDIM> > var,
+                        Pointer<Patch<NDIM> > patch,
+                        const double data_time,
+                        const bool initial_time = false,
+                        Pointer<PatchLevel<NDIM> > patch_level = Pointer<PatchLevel<NDIM> >(NULL));
```
Comment (boyceg) [inline examples/IBLevelSet/ex6/LevelSetInitialCondition.h:70]:
```suggestion
                        Pointer<PatchLevel<NDIM> > patch_level = nullptr);
```

File: examples/IBLevelSet/ex6/example.cpp [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+ * For restarted run, command line is:                                         *
+ *                                                                             *
+ *    executable <input file name> <restart directory> <restart number>        *
+ *                                                                             *
+ *******************************************************************************/
+int
+main(int argc, char* argv[])
+{
+    // Initialize PETSc, MPI, and SAMRAI.
+    PetscInitialize(&argc, &argv, NULL, NULL);
```
Comment (boyceg) [inline examples/IBLevelSet/ex6/example.cpp:144]:
```suggestion
    PetscInitialize(&argc, &argv, nullptr, nullptr);
```

File: examples/IBLevelSet/ex6/example.cpp [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+
+        RobinBcCoefStrategy<NDIM>* rho_bc_coef = NULL;
+        if (!(periodic_shift.min() > 0) && input_db->keyExists("DensityBcCoefs"))
+        {
+            rho_bc_coef = new muParserRobinBcCoefs(
+                "rho_bc_coef", app_initializer->getComponentDatabase("DensityBcCoefs"), grid_geometry);
+            navier_stokes_integrator->registerMassDensityBoundaryConditions(rho_bc_coef);
+        }
+
+        RobinBcCoefStrategy<NDIM>* mu_bc_coef = NULL;
```
Comment (boyceg) [inline examples/IBLevelSet/ex6/example.cpp:381]:
```suggestion
        RobinBcCoefStrategy<NDIM>* mu_bc_coef = nullptr;
```

File: examples/IBLevelSet/ex6/example.cpp [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+
+        // Close the logging streams.
+        if (SAMRAI_MPI::getRank() == 0)
+        {
+            vol_stream.close();
+            rbd_stream.close();
+            ft_stream.close();
+        }
+
+        // Delete dumb pointers.
```
Comment (boyceg) [inline examples/IBLevelSet/ex6/example.cpp:617]: Could we change "dumb" pointers to `std::unique_ptr`?

**Review by Raamkrishnan — CHANGES_REQUESTED** [review]
Looks good Kaustubh. Just two minor comments from my side.

File: examples/IBLevelSet/ex6/example.cpp (hunk 5 above)
Comment (Kaustubh-Khedkar) [reply examples/IBLevelSet/ex6/example.cpp:617]: This requires changing the interface of many hierarchy integrators, and none of the examples do that. Can you please point to an example if you know any that does this?

**Review by amneetb — APPROVED** [review]

**Review by Raamkrishnan — APPROVED** [review]

File: examples/IBLevelSet/ex6/example.cpp (hunk 5 above)
Comment (boyceg) [reply examples/IBLevelSet/ex6/example.cpp:616]: It shouldn't require changing any interfaces --- `unique_ptr<T>::get()` will return the underlying raw pointer `T*`.

(I think that we should try to move towards an API in which we use raw pointers or references to pass objects without passing ownership of the objects.)

Comment (Kaustubh-Khedkar) [reply examples/IBLevelSet/ex6/example.cpp]: I have made the requested changes. Please take a look and let me know.

**Comment by boyceg** [comment]
Looks good. Note that you don't need to initialize `std::unique_ptr<T>` to `nullptr` --- that will happen automatically.

File: examples/IBLevelSet/ex6/example.cpp [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+        if (periodic_shift.min() > 0)
+        {
+            for (unsigned int d = 0; d < NDIM; ++d)
+            {
+                u_bc_coefs[d] = nullptr;
+            }
+        }
+        else
+        {
+            for (unsigned int d = 0; d < NDIM; ++d)
```
Comment (boyceg) [inline examples/IBLevelSet/ex6/example.cpp:356]:
```suggestion
        if (periodic_shift.min() == 0)
        {
            for (unsigned int d = 0; d < NDIM; ++d)
```

File: examples/IBLevelSet/ex6/example.cpp [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+            navier_stokes_integrator->registerPhysicalBoundaryConditions({
+                u_bc_coefs[0].get(), u_bc_coefs[1].get()
+#if (NDIM == 3)
+                                         ,
+                    u_bc_coefs[2].get()
+#endif
+            });
+        }
+
+        std::unique_ptr<RobinBcCoefStrategy<NDIM> > rho_bc_coef = nullptr;
```
Comment (boyceg) [inline examples/IBLevelSet/ex6/example.cpp:378]:
```suggestion
        std::unique_ptr<RobinBcCoefStrategy<NDIM> > rho_bc_coef;
```

File: examples/IBLevelSet/ex6/example.cpp [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+        if (!(periodic_shift.min() > 0) && input_db->keyExists("DensityBcCoefs"))
+        {
+            rho_bc_coef = std::make_unique<muParserRobinBcCoefs>(
+                "rho_bc_coef", app_initializer->getComponentDatabase("DensityBcCoefs"), grid_geometry);
+            navier_stokes_integrator->registerMassDensityBoundaryConditions(rho_bc_coef.get());
+        }
+
+        std::unique_ptr<RobinBcCoefStrategy<NDIM> > mu_bc_coef = nullptr;
```
Comment (boyceg) [inline examples/IBLevelSet/ex6/example.cpp:386]:
```suggestion
        std::unique_ptr<RobinBcCoefStrategy<NDIM> > mu_bc_coef;
```

File: examples/IBLevelSet/ex6/example.cpp [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+        if (!(periodic_shift.min() > 0) && input_db->keyExists("ViscosityBcCoefs"))
+        {
+            mu_bc_coef = std::make_unique<muParserRobinBcCoefs>(
+                "mu_bc_coef", app_initializer->getComponentDatabase("ViscosityBcCoefs"), grid_geometry);
+            navier_stokes_integrator->registerViscosityBoundaryConditions(mu_bc_coef.get());
+        }
+
+        std::unique_ptr<RobinBcCoefStrategy<NDIM> > phi_bc_coef = nullptr;
```
Comment (boyceg) [inline examples/IBLevelSet/ex6/example.cpp:394]:
```suggestion
        std::unique_ptr<RobinBcCoefStrategy<NDIM> > phi_bc_coef;
```

**Comment by boyceg** [comment]
> Looks good. Note that you don't need to initialize `std::unique_ptr<T>` to `nullptr` --- that will happen automatically.

I think I marked the places where you don't need to explicitly initialize to `nullptr`.

File: examples/IBLevelSet/ex6/example.cpp (hunk 9 above)
Comment (Kaustubh-Khedkar) [reply examples/IBLevelSet/ex6/example.cpp]: I have made the requested changes. Please take a look.

**Review by boyceg — APPROVED** [review]

**Pull request merged**